libmamba/tests/src/util/test_iterator.cpp

```diff
@@ -0,0 +1,198 @@
+    {
+        return i > 10;
+    }
+
+    bool second_greater_than_10(const std::pair<std::string, int>& p)
+    {
+        return p.second > 10;
+    }
+
+    auto get_test_data()
+    {
+        std::vector<int> input_sequence = { 1, 12, 2, 3, 14, 4, 18, 20, 4 };
+        std::vector<int> result_sequence = { 12, 14, 18, 20 };
+        std::map<std::string, int> input_map =
+        {
+            {"a", 1},
+            {"b", 12},
+            {"c", 2},
+            {"f", 3},
+            {"x", 14},
+            {"g", 4},
+            {"d", 18},
+            {"l", 20},
+            {"n", 4}
+        };
+        std::map<std::string, int> result_map =
+        {
+            {"b", 12},
+            {"x", 14},
+            {"d", 18},
+            {"l", 20}
+        };
+
+        return std::make_tuple(input_sequence, result_sequence, input_map, result_map);
+
+    }
+}
+
+TEST_SUITE("filter_iterator")
+{
+    TEST_CASE("forward_iterator_api")
+    {
+        auto [input_seq, res_seq, input_map, res_map] = get_test_data();
+        {
+            auto iter = filter_begin(greater_than_10, input_seq);
+            auto iter_end = filter_end(greater_than_10, input_seq);
+
+            auto citer = cfilter_begin(greater_than_10, input_seq);
+            auto citer_end = cfilter_end(greater_than_10, input_seq);
+
+            auto res_iter = res_seq.begin();
+
+            auto iter2 = iter, iter3 = iter;
+
+            for (size_t i = 0; i < res_seq.size(); ++i)
+            {
+                CHECK_EQ(*iter, *res_iter);
+                CHECK_EQ(*iter, *citer);
+                ++iter, ++citer, ++res_iter;
+            }
+            CHECK_EQ(iter, iter_end);
+            CHECK_EQ(citer, citer_end);
+            CHECK(iter == iter_end);
+
+            CHECK_EQ(iter2.operator->(), &input_seq[1]);
+            CHECK(iter2++ != ++iter3);
+        }
+        {
+            auto iter = filter_begin(second_greater_than_10, input_map);
+            auto iter_end = filter_end(second_greater_than_10, input_map);
+
+            auto citer = cfilter_begin(second_greater_than_10, input_map);
+            auto citer_end = cfilter_end(second_greater_than_10, input_map);
+
+            auto res_iter = res_map.begin();
+
+            auto iter2 = iter, iter3 = iter;
+
+            for (size_t i = 0; i < res_map.size(); ++i)
+            {
+                CHECK_EQ(*iter, *res_iter);
+                CHECK_EQ(*iter, *citer);
+                ++iter, ++citer, ++res_iter;
+            }
+            CHECK_EQ(iter, iter_end);
+            CHECK_EQ(citer, citer_end);
+            CHECK(iter == iter_end);
+            
+            CHECK_EQ(iter2->second, res_map.begin()->second);
+            CHECK(iter2++ != ++iter3);
+        }
+    }
+
+    TEST_CASE("bidirectional_iterator_api")
+    {
+        auto [input_seq, res_seq, input_map, res_map] = get_test_data();
+        {
+            auto iter = filter_begin(greater_than_10, input_seq);
```

Comment:
`const`. 


`const` everywhere! 🤡

Reply:
Not sure to get it, `iter` is changed after, and I want to test non const iterators so the sequances / maps cannot be const.